### ISSUE TITLE
🐛 noUnusedImports supports custom jsxFactory

### ISSUE THREAD

**Issue by ahaoboy** [biomejs/biome#3682]
### Environment information

```block
CLI:
  Version:                      1.8.3
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           windows

Environment:
  BIOME_LOG_DIR:                unset
  NO_COLOR:                     unset
  TERM:                         "xterm-256color"
  JS_RUNTIME_VERSION:           "v22.6.0"
  JS_RUNTIME_NAME:              "node"
  NODE_PACKAGE_MANAGER:         "pnpm/9.6.0"

Biome Configuration:
  Status:                       Loaded successfully
  Formatter disabled:           false
  Linter disabled:              false
  Organize imports disabled:    false
  VCS disabled:                 true

Workspace:
  Open Documents:               0
```

### What happened?

https://github.com/ahaoboy/biome-react-jsxFactory

When using a custom renderer, after turning on noUnusedImports, the jsxFactory field configured in tsconfig is not correctly recognized

For example, in the following code, only variable b should not be used.
However, biome will mark both $ and b as unused.
```json
    "jsx": "react",
    "jsxFactory": "$",
    "jsxFragmentFactory": "$"
```
```tsx
import { $, b } from "./react"
function App() {
  return <div />
}
export default <App />
```

### Expected result

biome can handle custom jsxFactory correctly
For the above code, only variable b should not be used.

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct

**Comment by ahaoboy** [comment]
Is it possible to add a separate configuration? For a very large monorepo, there may not be a tsconfig in the root directory
```
  "javascript": {
    "jsxRuntime": {
      "jsx": "react",
      "jsxFactory": "$",
      "jsxFragmentFactory": "$",
    },
  },
```

**Comment by ematipico** [comment]
PRs are welcome if anyone wants to help

**Comment by h-a-n-a** [comment]
I'm new to biome.js but I'm very glad to help if you can bear with me taking a bit longer to fix this ❤️

**Comment by h-a-n-a** [comment]
I was wondering if we can support:

```json
{
  "javascript": {
    "jsxRuntime": "reactClassic",
    "jsxFactory": "$",
    "jsxFragmentFactory": "$"
  }
}
```
Note: `jsxFactory` or `jsxFragmentFactory` will only be used when `jsxRuntime` is set to `"react"`. 

If we were to support struct type of `jsxRuntime`, this would cause a breaking change to config file. Otherwise, we may need to support either type in deserialization for `jsxRuntime`, which could probably make JSON schema hard to maintain as well. However, flattening these options do not cause any of the problems above.

**Comment by ahaoboy** [comment]
Considering that different subprojects in monorepo may have different configurations, it would be nice to read the configurations from the tsconfig files of the subprojects, but there is extends field in tsconfig, and it is not clear if biome supports it.

**Comment by ematipico** [comment]
@ahaoboy if a project doesn't use TypeScript, how does someone provide those values?

**Comment by ahaoboy** [comment]
> @ahaoboy if a project doesn't use TypeScript, how does someone provide those values?

Yes, this is why I want to add a configuration. Can we use the default configuration for projects without tsconfig?
```
import React from react
```

**Comment by ematipico** [comment]
@h-a-n-a Your proposal sounds reasonable. I suppose we could add an ad-hoc validation that emits an error/warning in case of incorrect configuration. Feel free to start the implementation as you see fit

@ahaoboy thank you for your answer

**Comment by h-a-n-a** [comment]
@ematipico Hi! I encountered some problem.

## Background

 I dug into the codebase. I found out the original implementation is a little bit simple:
https://github.com/arendjr/biome/blob/67d03778769c9b944b03575032f4bdbcabec27c9/crates/biome_js_analyze/src/lint/correctness/no_unused_imports.rs#L99-L103
It completely ignored the case if `jsxRuntime` is set to `ReactClassic` and if it's a global react import statement.

The following module does not complain any error even though the identifier binding `React` is not being used:

```js
import React from "react"
```

The other issue I encountered is if `jsxRuntime` is set to `ReactClassic`, and with the following code:

```jsx
function App() {
  return <><div>abc</div></>
}

export default <App />
```

Biome does not complain anything (note that `React` is not being imported in this case). However, as per the `jsxRuntime` section of the document: 
> Corresponds to the react value for the jsx option in TypeScript’s [tsconfig.json](https://www.typescriptlang.org/tsconfig#jsx)
And at the same time, TypeScript compiler does complain as follows:

```
Cannot find name 'React'.
```

So I believe there does have something to improve and I want to focus on this issue first.

## Possible solution

I would like to improve the binding. 
For example, for `jsxRuntime` being set to `ReactClassic` and `jsxFactory` being set to `h`, both of the following examples should not complain:

```js
// 1
const h = () => {}
const Element = <div></div>

// 2
import { h } from "preact"
const Element = <div></div>
```

Also, jsx runtime should be scope aware, with the same configuration above, the following example should complain:

```jsx
function App() {
  const h = () => {}
  return <div /> // 1
}

export default <App /> // 2 👈 Cannot find binding `h`
```

Note 1 should not complain as `h` is being defined locally, however 2 does not have any identifier binding defined.

## Question

I took a look at the `SemanticEventExtractor` and I was thinking of every time a `SyntaxKind::JSX_ELEMENT` is visited, a reference should be added to the scope so that I can get the reference in `no_used_imports` with `binding.all_references` without changing the plugin code. However, maybe I'm taking this wrong, because of every `BindingName` and `Reference` is from the raw node in roman, there's no place to add a synthetic binding.

Would you please help me with a little guidance? 
Thanks for reading and being patient with me!

**Comment by ematipico** [comment]
>  It completely ignored the case if jsxRuntime is set to ReactClassic and if it's a global react import statement.

I believe it's intended, because in "classic" react, the old way, the global `React` was mandatory. This code should pass, because the old version of babel JSX transformer **for react** requires it:

```js
import React from "react"

export function App() {
	return <p>hello</p>
}
```

However, a more modern version of the babel JSX transformer doesn't require it anymore. In lue to that, we decided to ignore the import in case we have a classic (old) transformer.

> So I believe there does have something to improve and I want to focus on this issue first.

Please do. Consider that, at the moment, we can't read the `tsconfig.json` yet. It's coming soon. 
___
Regarding your question, I believe the semantic model already does that 🤔

**Comment by h-a-n-a** [comment]
> Regarding your question, I believe the semantic model already does that 🤔

AFAIK, they only do bindings for cases like: 

```js
const App = () => {}
<App /> // App usage
```

But instead, I'd like to add a binding to `jsxFactory` and `jsxFragmentFactory`. 

Ref:  https://github.com/arendjr/biome/blob/01a80a47afae76d80f847971a979b3315d3eda67/crates/biome_js_semantic/src/events.rs#L747
